dgraph/cmd/raft-migrate/run.go

```diff
@@ -82,6 +82,51 @@ func parseAndConvertSnapshot(snap *raftpb.Snapshot) {
 	x.Check(err)
 }
 
+func shouldMigrate(oldWal *raftwal.DiskStorage, oldEntries []raftpb.Entry) bool {
+	isOldFormat := func(entry raftpb.Entry) bool {
+		if entry.Data[0] == 0 {
+			return false
+		}
+		return true
+	}
+
+	groupID := oldWal.Uint(raftwal.GroupId)
+	if groupID == 0 {
+		// For zero we need to verify both the Snapshot as well as raft-entries.
+		snapshot, err := oldWal.Snapshot()
+		x.Checkf(err, "failed to read snaphot %s", err)
+		var ms pb.MembershipState
+		if err := ms.Unmarshal(snapshot.Data); err != nil {
+			// We are unmarshalling the ZeroSnapshot into MembershipState. If this fails, this means
+			// we are on newer version.
+			fmt.Println("Zero snapshot unmarshalling failed => New format")
+			return false
+		}
+		for _, entry := range oldEntries {
+			// Raft commits an empty entry on becoming leader.
+			if entry.Type == raftpb.EntryConfChange || len(entry.Data) == 0 {
+				continue
+			}
+			// In new format, we prepend key to entry.Data in first 8 bytes. First 4 bytes out of
+			// those contain a node ID. If we assume node ID to be less than 1<<24. Then, if
+			// entry.Data[0] = 0, we can conclude that we are on new format.
```

Comment:
rather than this, can we try Unmarshalling `entry.Data[8:]`? if it fails then we are on oldWal.

Reply:
> if it fails then we are on oldWal

But that won't be conclusive. Mostly it would return an error but if you're unlucky, you might be able to unmarshal it.